// node-opcua-nodeset-ua/dist/enum_identity_criteria.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumIdentityCriteria = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |IdentityCriteriaType                              |
 * | isAbstract|false                                             |
 */
var EnumIdentityCriteria;
(function (EnumIdentityCriteria) {
    EnumIdentityCriteria[EnumIdentityCriteria["UserName"] = 1] = "UserName";
    EnumIdentityCriteria[EnumIdentityCriteria["Thumbprint"] = 2] = "Thumbprint";
    EnumIdentityCriteria[EnumIdentityCriteria["Role"] = 3] = "Role";
    EnumIdentityCriteria[EnumIdentityCriteria["GroupId"] = 4] = "GroupId";
    EnumIdentityCriteria[EnumIdentityCriteria["Anonymous"] = 5] = "Anonymous";
    EnumIdentityCriteria[EnumIdentityCriteria["AuthenticatedUser"] = 6] = "AuthenticatedUser";
    EnumIdentityCriteria[EnumIdentityCriteria["Application"] = 7] = "Application";
})(EnumIdentityCriteria = exports.EnumIdentityCriteria || (exports.EnumIdentityCriteria = {}));
