// node-opcua-factory/dist/get_standard_data_type_factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dump = exports.hasConstructor = exports.getConstructor = exports.hasStructuredType = exports.getStructureTypeConstructor = exports.getStandardDataTypeFactory = void 0;
const datatype_factory_1 = require("./datatype_factory");
let globalFactory;
function getStandardDataTypeFactory() {
    if (!globalFactory) {
        globalFactory = new datatype_factory_1.DataTypeFactory([]);
        globalFactory.targetNamespace = "http://opcfoundation.org/UA/";
    }
    return globalFactory;
}
exports.getStandardDataTypeFactory = getStandardDataTypeFactory;
function getStructureTypeConstructor(typeName) {
    return getStandardDataTypeFactory().getStructureTypeConstructor(typeName);
}
exports.getStructureTypeConstructor = getStructureTypeConstructor;
function hasStructuredType(typeName) {
    return getStandardDataTypeFactory().hasStructuredType(typeName);
}
exports.hasStructuredType = hasStructuredType;
function getConstructor(binaryEncodingNodeId) {
    return getStandardDataTypeFactory().getConstructor(binaryEncodingNodeId);
}
exports.getConstructor = getConstructor;
function hasConstructor(binaryEncodingNodeId) {
    return getStandardDataTypeFactory().hasConstructor(binaryEncodingNodeId);
}
exports.hasConstructor = hasConstructor;
/* istanbul ignore next */
function dump() {
    getStandardDataTypeFactory().dump();
}
exports.dump = dump;
